File: ui/NSWindow+Fade [osx]/NSWindow+Fade.m
//
//  NSWindow+Fade.m
//  GoDark
//
//  Created by Dominik Pich on 09.12.12.
//  Copyright (c) 2012 info.pich. All rights reserved.
//

#import "NSWindow+Fade.h"

#define kWindowAnimationDuration 0.1f

@implementation NSWindow (Fade)

- (void)fadeInWithDuration:(NSNumber*)d
{
    [self setAlphaValue:0.f];
    [self makeKeyAndOrderFront:nil];
    [NSAnimationContext beginGrouping];
    [[NSAnimationContext currentContext] setDuration:d.floatValue];
    [[self animator] setAlphaValue:1.f];
    [NSAnimationContext endGrouping];
}

- (IBAction)fadeIn:(id)sender
{
    [self fadeInWithDuration:@(kWindowAnimationDuration)];
}

- (void)fadeOutWithDuration:(NSNumber*)d
{
    [NSAnimationContext beginGrouping];
    __block NSWindow *bself = self;
    [[NSAnimationContext currentContext] setDuration:d.floatValue];
    [[NSAnimationContext currentContext] setCompletionHandler:^{
        [bself orderOut:nil];
        [bself setAlphaValue:1.f];
    }];
    [[self animator] setAlphaValue:0.f];
    [NSAnimationContext endGrouping];
}

- (IBAction)fadeOut:(id)sender
{
    [self fadeOutWithDuration:@(kWindowAnimationDuration)];
}

@end

File: swift/ddutils-osx/ui/NSWindow+Fade [osx]/NSWindow+Fade.swift
//
//  NSWindow+Fade.swift
//
//  Created by Dominik Pich on 4/27/16.
//  Copyright © 2016 Dominik Pich. All rights reserved.
//

import Cocoa

let kWindowFadeAnimationDuration = 0.1

extension NSWindow /*(Fade)*/ {
    func fadeInWithDuration(duration: NSTimeInterval) {
        alphaValue = 0
        makeKeyAndOrderFront(nil)
        
        NSAnimationContext.beginGrouping()
        NSAnimationContext.currentContext().duration = duration
        animator().alphaValue = 1
        NSAnimationContext.endGrouping()
    }
    
    @IBAction func fadeIn(sender: AnyObject!) {
        fadeInWithDuration(kWindowFadeAnimationDuration)
    }
    
    func fadeOutWithDuration(duration: NSTimeInterval) {
        NSAnimationContext.beginGrouping()
        NSAnimationContext.currentContext().duration = duration
        NSAnimationContext.currentContext().completionHandler = {
            self.orderOut(nil)
            self.alphaValue = 1
        }
        animator().alphaValue = 0
        NSAnimationContext.endGrouping()
    }
    
    @IBAction func fadeOut(sender: AnyObject!) {
        fadeOutWithDuration(kWindowFadeAnimationDuration)
    }
}